### ISSUE TITLE
Swipe to dismiss

### ISSUE THREAD

**Issue by PauloxD** [SCENEE/FloatingPanel#8]
Hi just started using this library, and it is really clean, so well done on that.

Jut wondering how to go about performing a swipe to dismiss style move?

Thanks

**Comment by Yonodactyl** [comment]
Hey @PauloxD 

Hope you are doing alright today.
Looking through the source code for the framework, it seems there is a func that can be called to dismiss the panel.

```swift
removePanelFromParent(animated: Bool, completion: (() -> Void)?
```

So once you have established a FloatingPanel variable, you can just call that function a little something like: 
```swift
fpc.removePanelFromParent(animated: true, completion: nil)
```
This just performs a weak dismissal of the floatingPanel.

**Comment by PauloxD** [comment]
Firstly, thank you for replying so quickly.

Yeah that makes sense, and I use that upon completing the form that I am using this library for. My issue is identifying that a swipe action has happened on the child controller, or a tap has happened on the parent controller. Does that make sense?

So, normally I would use the scroll delegate or touch delegate to calculate if the movement is a downward gesture. 

Once again, thanks for replying

**Comment by SCENEE** [comment]
@PauloxD, I really appreciate your comments,

> a tap has happened on the parent controller. 

On this case, this code might be your help.
https://github.com/SCENEE/FloatingPanel/blob/master/Examples/Samples/Sources/ViewController.swift#L127

> a swipe action has happened on the child controller
>  I would use the scroll delegate or touch delegate to calculate if the movement is a downward gesture.

Now this library doesn't case of your use case.  But I will investigate it's possible or not😀

You can still use delegate methods of a tracked scroll view in the child controller. Please note that you set a delegate object to a scroll view before tracking it.
https://github.com/SCENEE/FloatingPanel/blob/master/Framework/Sources/FloatingPanelController.swift#L261

**Comment by Urkman** [comment]
I'm also interested in some kind of Swipe to dismiss...
Because I don't want to display the Panel all the time. It should only appear if the user needs it and he should also be able to close it, after he uses it...

Would be grate to a example for this :)

**Comment by SCENEE** [comment]
@PauloxD, @Urkman 

Thank you for your patient. Could you please try 'add-removable-interaction' branch?
You can see a sample of a removable panel("Show Removable Panel") in Samples app.

https://github.com/SCENEE/FloatingPanel/pull/25/files#diff-b733e6931fa8657f8bc9ba741b879955R67

**Comment by Urkman** [comment]
Just Xcode 10.1 and iPhone XS Max in Simulator...

**Comment by Urkman** [comment]
I did some more tests... And sometime it works, but most of time not :(
It works better, when I first swipe up and then swipe down to dismiss...

**Comment by SCENEE** [comment]
Thank you for your investigation. I'm going to investigate why it's unstable so I reopen this issue.

**Comment by SCENEE** [comment]
@Urkman,
I think I fix this issue on 'fix-removal-interaction' branch.  Could you try it?
Thank you for your cooperation.

**Comment by Urkman** [comment]
Still not working for me :(
I try to grab the handle an pull it down. But it always slides back up...

**Comment by SCENEE** [comment]
I got it. The implementation you tried removed a floating panel just on swiping down with over a velocity. So I add a floating panel's coordinate Y condition to remove it and update the branch.
Now "Removable Pane" should behave as you like.

**Comment by Urkman** [comment]
Great... I will test it as soon i‘m in the office

**Comment by Urkman** [comment]
Works great :)

Thanks...

**Comment by SCENEE** [comment]
Thanks, too 👍

**Comment by Urkman** [comment]
Two more :(

1.) Orientation change not handled properly with "Removable panel"
2.) After "Swipe to dismiss" Panel is not removed.

Reproducible in Sample:

Working:
1.) click to show "Removable panel"
2.) Click Close Button
3.) Change Orientation
4.) Everything is fine (No panel shown)

Problem "Orientation Change":
1.) click to show "Removable panel"
2.) Change Orientation
3.) See, that Panel is not shown correctly 
4.) Change Orientation back
5.) Panel is only show in lower third of the screen
6.) Try to swipe panel up to "half.
7.) Panel is dismissed

Problem "Panel not removed after Swipe to dismiss"
1.) click to show "Removable panel"
2.) swipe to dismiss panel
3.) Panel is dismissed
4.) Change Orientation
5.) Panel is show again :(

I hope my descriptions can help you to find the problem...

Thanks again for your great work. I will definitely use this in my next project :)

**Comment by Urkman** [comment]
And one more question...

Is there a delegate, that is called, when a panel is dismissed?

**Comment by Urkman** [comment]
I found the delegate "floatingPanelDidEndRemove"...
With this I can remove the Controller from it's parent :)
Problem "Panel not removed after Swipe to dismiss" can be handled with this :)

Leaves the "Orientation Change" Problem open...

**Comment by SCENEE** [comment]
I could find some additional bugs on orientation change. Thanks a lot.
I hope I would fix the above problem on PR #37. Could you please try PR #37 (`fix-orientation-change-on-removal` branch)?

**Comment by Urkman** [comment]
Lock better, when I use fixed panel hights... BUT...

I want the panel to take exactly half the space of the screen.
So I use this to calculate the hight in my layout files:

For portrait:
`let hight = max(UIScreen.main.bounds.size.width, UIScreen.main.bounds.size.height) / 2`

For landscape:
`let hight = min(UIScreen.main.bounds.size.width, UIScreen.main.bounds.size.height)`

And then I get a crash on changing orientation from landscape to portrait:
> 2018-11-07 13:00:02.643397+0100 name[2353:535426] Assertion failed: Invalid insets: file /path/Pods/FloatingPanel/Framework/Sources/FloatingPanelLayout.swift, line 288

**Comment by Urkman** [comment]
To be more exact here:
`    func insetFor(position: FloatingPanelPosition) -> CGFloat? {
        switch position {
        case .full: return 16.0
        case .half: return 380
        default: return nil
        }
    }
`

With this in my Portrait Layout File the crashes on my iPhone XS Max on location change from landscape to portrait

**Comment by SCENEE** [comment]
A crash happened by an assertion failure to your layout because it's invalid for landscape.

But it's confusing because it's a wrong time for this library to raise the assertion failed. It should be raised on changing orientation from portrait to landscape. This issue was fixed on `6e85afa`.

If you configure an appropriate a layout for landscape, it can work. [support-your-landscape-layout](https://github.com/SCENEE/FloatingPanel#support-your-landscape-layout) section can help you.

**Comment by Urkman** [comment]
Sorry, I don't get it :(
Why should my Layout for Landscape be invalid?

It opens fine in Landscape and it open fine in Portrait...
The problem raises on orientation change. 

If I remove these asserts:
`assert(middleY > topY, "Invalid insets")
assert(bottomY > topY, "Invalid insets")
`

**Comment by SCENEE** [comment]
> Why should my Layout for Landscape be invalid?

Hmmm...🤔, this is because a floating  panel is displayed as the below screen shot by `func insetFor(position: FloatingPanelPosition) -> CGFloat? { switch position { case .full: return 16.0 case .half: return 380 default: return nil } }`.
![simulator screen shot - iphone xs max - 2018-11-09 at 09 00 36](https://user-images.githubusercontent.com/899487/48234702-ff0be200-e3fd-11e8-8e16-b23bd50b3bdb.png)

I guess `var supportedPositions: Set<FloatingPanelPosition> { return [.full, .half] }`.
Then the layout introduces `topY: 104.0` and `middleY:  0.0` / `bottomY: 0.0` on iPhone XS Max.
```
topY = 88.0 + 16.0
middleY = 414pt(height) - 380pt + 34.0pt(safe area bottom) = 0pt
bottomY = 414pt(height) - 380pt + 34.0pt(safe area bottom) = 0pt
```

**Comment by Urkman** [comment]
I try to make a small example to show the orientation change problem...

**Comment by SCENEE** [comment]
Thanks, @Urkman. Could you please open a new issue?
I'd like to discuss the orientation problem in it. because I think the problem is different from this issue.

**Comment by trungnguyen1791** [comment]
@SCENEE I'm facing an issue that i can't swipe to show delete action on tableview in FloatingPanel.
When i disable this line flc.track(scrollView: vc.tableView), the swipe gesture on tableview works again. But without it i can't swipe to close panel.

For your information:
- current i'm using/showing panel in a tabbar controller with navigation controller also ( not standalone tabbar controller like in your sample).
- using version 1.2.0.

**Comment by SCENEE** [comment]
Thanks, @trungnguyen1791! I will take care of it.

**Comment by SCENEE** [comment]
The original issue has already resolved on v1.2.0.
I would like to track the issue reported by @trungnguyen1791 on PR #56. So I close this.
